src/Oscoin/Storage/Block/SQLite/Tx.hs

```diff
@@ -0,0 +1,162 @@
+        txRowAuthor    <- field
+
+        pure TxRow{..}
+
+
+newtype TxRowDecodeError = TxRowDecodeError String
+    deriving (Show, Eq)
+
+instance Exception TxRowDecodeError where
+    displayException (TxRowDecodeError msg) =
+        "Failed to decode transaction row: " <> msg
+
+toTxRow ::
+    ( StorableTx c
+    ) => Tx c -> TxRow c
+toTxRow tx@Tx'{..} = TxRow
+    { txRowHash      = fromHashed $ hash tx
+    , txRowNetwork   = fromIntegral $ fromEnum txNetwork
+    , txRowSignature = txSignature
+    , txRowMessages  = LBS.toStrict $ serialise (txMessages txPayload)
+    , txRowNonce     = txNonce txPayload
+    , txRowFee       = txFee txPayload
+    , txRowBurn      = txBurn txPayload
+    , txRowAuthor    = txAuthor txPayload
+    }
+
+fromTxRow :: StorableTx c => TxRow c -> Either TxRowDecodeError (Tx c)
+fromTxRow TxRow{..} = first toDecodeError $ do
+    txNetwork   <- pure $ toEnum (fromIntegral txRowNetwork)
+    txSignature <- pure txRowSignature
+    txMessages  <- deserialiseOrFail $ LBS.fromStrict txRowMessages
+    txNonce     <- pure txRowNonce
+    txFee       <- pure txRowFee
+    txBurn      <- pure txRowBurn
+    txAuthor    <- pure txRowAuthor
+
+    pure Tx'
+        { txPayload = TxPayload
+            { txMessages
+            , txNonce
+            , txFee
+            , txBurn
+            , txAuthor
+            }
+        , txNetwork
+        , txSignature
+        }
+  where
+    toDecodeError :: DeserialiseFailure -> TxRowDecodeError
+    toDecodeError = TxRowDecodeError . displayException
+
+getTx :: forall c. (StorableTx c, Typeable c) => Sql.Connection -> Hash c -> IO (Maybe (Tx c))
+getTx conn hsh = do
+    rows <- Sql.query conn
+        [sql| SELECT hash, network, signature, messages, nonce, fee, burn, author
+                FROM transactions
+               WHERE hash = ? |] (Only hsh)
+    forM (listToMaybe rows) (fromTxRowThrow @c)
+
+
+storeTxs :: forall c. (StorableTx c) => Sql.Connection -> Hash c -> [Tx c] -> IO ()
+storeTxs conn blockHash txs =
+    -- Nb. To relate transactions with blocks, we store an extra block hash field
+    -- for each row in the transactions table.
+    Sql.executeMany conn
+        [sql| INSERT INTO transactions  (hash, network, signature, messages, nonce, fee, burn, author, blockhash)
+              VALUES                    (?, ?, ?, ?, ?, ?, ?, ?, ?) |] (map toTxRowValues txs)
+  where
+    toTxRowValues tx =
+        let TxRow{..} = toTxRow @c tx
+        in
+            [ toField txRowHash
+            , toField txRowNetwork
+            , toField txRowSignature
+            , toField txRowMessages
+            , toField txRowNonce
+            , toField txRowFee
+            , toField txRowBurn
+            , toField txRowAuthor
+            , toField blockHash
+            ]
+
+-- | Get the transactions belonging to a block.
+getBlockTxs :: forall c. (StorableTx c, Typeable c) => Connection -> BlockHash c -> IO [Tx c]
+getBlockTxs conn h = do
```

Comment:
I find `getTx/getBlockTxs` a bit weird from the API perspective -- the latter doesn't return a `Maybe`, whereas the former does. Now, I do appreciate that probably in the latter case is hard (if not impossible) to distinguish the failure from the fact we return an empty list because there are no transactions for a block, but still my OCD wonder whether we could improve this somehow. It doesn't matter too much, but it would still be nice to think about it 😉

Reply:
I think renaming the former to `lookupTx` is a first step to distinguishing them. I also just copied these over from the old SQLite.Transaction module, so didn't really think about it.